lib/fluent/plugin/out_logdna.rb

```diff
@@ -66,11 +65,33 @@ def chunk_to_body(chunk)
     end
 
     def gather_line_data(tag, time, record)
+      # record is a fluentd record which looks close to what we're going to send to LogDNA.
+      # We need to promote a couple of fields (providing some defaults), and then send
+      # everything else as-is.
       line = {
         level: record['level'] || record['severity'] || tag.nil? ? "INFO" : tag.split('.').last,
         timestamp: time,
-        line: record.to_json
       }
+
+      # The default JSON library will raise an exception if encoding fails.
+      # The HTTPrb library will attempt to encode all json content into UTF-8.
+      # Therefore we need to clense invalid UTF-8 that may be in your message
+      # (say because you're logging raw user input or binary files...). We'll
+      # also set encoding_error to true so that you know the message has been
+      # lossily altered.
+      begin
+        record['message'].encode!("UTF-8")
```

Comment:
is there a reason not to just do the first encoding pass with replacement other than marking the error condition? i'm curious if there's another way to get "yes i did replace some characters" out of `encode!`

Reply:
reading the ruby doc on `encode!` and `encode` https://ruby-doc.org/core-2.1.3/String.html#method-i-encode , I don't think there's a way to replace characters and tell if the replacement happens at one step. Looks like the `Raise` error is the only way to tell whether sth has been replaced here.

Reply:
I found https://ruby-doc.org/core-2.1.0/String.html#method-i-scrub but don't see a good way to use it here since we're not already in UTF-8. I think what you've got will be fine.